sdk/dynamicconfig/metric/controller/push/period_matcher.go

```diff
@@ -51,26 +52,48 @@ func (matcher *PeriodMatcher) MarkStart(startTime time.Time) {
 // ApplySchedules sets the schedules that a PeriodMatcher consults when
 // constructing a Rule. After processing the schedules, ApplySchedules returns
 // the optimal period with which a controller should run a collection sweep.
+// If a period of 0 is returned, then metric collection should be halted.
 //
 // This function may be called concurrently.
-func (matcher *PeriodMatcher) ApplySchedules(sched []*pb.MetricConfigResponse_Schedule) time.Duration {
+func (matcher *PeriodMatcher) ApplySchedules(sched []*pb.MetricConfigResponse_Schedule) (time.Duration, error) {
+	if err := validate(sched); err != nil {
+		return 0, err
+	}
+
+	exportPeriod := getExportPeriod(sched)
+
 	matcher.m.Lock()
 	matcher.sched = sched
 	matcher.metrics = make(map[string]*collectData)
 	matcher.m.Unlock()
 
-	return getExportPeriod(matcher.sched)
+	return exportPeriod, nil
+}
+
+func validate(sched []*pb.MetricConfigResponse_Schedule) error {
+	if len(sched) == 0 {
+		return errors.New("no schedules")
+	}
+
+	for _, schedule := range sched {
+		if schedule.PeriodSec < 0 {
+			return errors.New("periods must be positive")
+		}
+	}
+
+	return nil
 }
 
-// TODO: handle explicit zeros
 func getExportPeriod(sched []*pb.MetricConfigResponse_Schedule) time.Duration {
```

Comment:
What if all periods in sched are zero?

Reply:
Great question, the bulk is in the above answer, but in brief, it disables metric collection, since a zero period implies "don't collect this metric."